modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcServer.java

```diff
@@ -175,13 +179,37 @@ else if (prc.executor() != null) {
             RpcProcessor<NetworkMessage> finalPrc = prc;
 
             try {
-                executor.execute(() -> finalPrc.handleRequest(new NetworkRpcContext(executor, sender, correlationId), message));
+                if (shouldSwitchToRequestsExecutor()) {
+                    executor.execute(() -> finalPrc.handleRequest(new NetworkRpcContext(executor, sender, correlationId), message));
+                } else {
+                    finalPrc.handleRequest(new NetworkRpcContext(executor, sender, correlationId), message);
+                }
             } catch (RejectedExecutionException e) {
                 // The rejection is ok if an executor has been stopped, otherwise it shouldn't happen.
                 LOG.warn("A request execution was rejected [sender={} req={} reason={}]", sender, S.toString(message), e.getMessage());
             }
         }
 
+        private boolean shouldSwitchToRequestsExecutor() {
```

Comment:
This method differs from the corresponding method in ReplicaManager in a tiny way (this method does not check for 3rd permission). Would it make sense to extract a common method and parameterize it with the set of permissions to check, to avoid duplication?